src/cryptonote_core/tx_pool.cpp

```diff
@@ -227,38 +227,38 @@ namespace cryptonote
 
     crypto::hash max_used_block_id = null_hash;
     uint64_t max_used_block_height = 0;
-    cryptonote::txpool_tx_meta_t meta;
-    bool ch_inp_res = check_tx_inputs([&tx]()->cryptonote::transaction&{ return tx; }, id, max_used_block_height, max_used_block_id, tvc, kept_by_block);
+    cryptonote::txpool_tx_meta_t meta{};
+    bool ch_inp_res = check_tx_inputs([&tx]()->cryptonote::transaction&{ return tx; }, id, max_used_block_height, max_used_block_id, tvc, tx_relay == relay_method::block);
     if(!ch_inp_res)
     {
       // if the transaction was valid before (kept_by_block), then it
       // may become valid again, so ignore the failed inputs check.
-      if(kept_by_block)
+      if(tx_relay == relay_method::block)
       {
         meta.weight = tx_weight;
         meta.fee = fee;
         meta.max_used_block_id = null_hash;
         meta.max_used_block_height = 0;
         meta.last_failed_height = 0;
         meta.last_failed_id = null_hash;
-        meta.kept_by_block = kept_by_block;
         meta.receive_time = receive_time;
         meta.last_relayed_time = time(NULL);
         meta.relayed = relayed;
-        meta.do_not_relay = do_not_relay;
+        meta.set_relay_method(tx_relay);
         meta.double_spend_seen = have_tx_keyimges_as_spent(tx);
         meta.pruned = tx.pruned;
         meta.bf_padding = 0;
         memset(meta.padding, 0, sizeof(meta.padding));
         try
         {
-          if (kept_by_block)
+          if (tx_relay == relay_method::block)
             m_parsed_tx_cache.insert(std::make_pair(id, tx));
           CRITICAL_REGION_LOCAL1(m_blockchain);
           LockedTXN lock(m_blockchain);
-          m_blockchain.add_txpool_tx(id, blob, meta);
-          if (!insert_key_images(tx, id, kept_by_block))
+          if (!insert_key_images(tx, id, tx_relay))
             return false;
+
+          m_blockchain.add_txpool_tx(id, blob, meta);
```

Comment:
Why was this reordered ? The ordering means that key images get inserted only if the DB write succeeds. Otherwise, you need to remove key images on DB commit failure since hte key images are in memory, not the db txn.

Reply:
This call changes the status of the tx in-place. A stem tx is converted to a fluff tx, etc. This also updates the timestamp (verified in new core tests) so that information isn't leaked either. Everything is replaced.

This had to be after `insert_key_images` because that call is doing a DB lookup to find its current status.